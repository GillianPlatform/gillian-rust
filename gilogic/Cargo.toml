[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }