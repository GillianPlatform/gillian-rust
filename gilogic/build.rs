fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
