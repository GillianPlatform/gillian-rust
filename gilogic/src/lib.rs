#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
