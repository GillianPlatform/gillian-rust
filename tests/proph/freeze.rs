//@ check-pass
extern crate creusillian;
extern crate gilogic;

use gilogic::{
    macros::*,
    prophecies::{Ownable, Prophecised, Prophecy},
};

struct A {
    x: u64,
    y: u64,
}

#[with_freeze_lemma(
    lemma_name = freeze_x,
    predicate_name = a_ref_mut_x,
    frozen_variables = [x]
)]
impl Ownable for A {
    type RepresentationTy = u64;

    #[predicate]
    fn own(self, model: Self::RepresentationTy) {
        assertion!(|x: u64, y| (self == A { x, y }) * (model == x) * (y == x + 1))
    }
}
