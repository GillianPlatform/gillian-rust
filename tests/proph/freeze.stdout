#begin_init_data
{"A":["Struct","ReprRust",[["x",["Scalar","u64"]],["y",["Scalar","u64"]]]]}
#end_init_data

import "i__binop.gil", "i__lang.gil", "i__std_shims.gil", "i__prophecies.gil", "i__std_shims_pcy.gil", "i__lang_pcy.gil";

axiomatic spec "gilogic::prophecies::freeze_params::<(u64,), A>"(pLft_0, p)
[[ "<&mut A as gilogic::prophecies::Ownable>::own"(#pLft_0, #p, #m)
  * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj)
  * (#p == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }})
  * (p == #p)
  * (pLft_0 == #pLft_0) ]]
[[ "<A as gilogic::prophecies::FrozenOwn<(u64,)>>::mut_ref_own_frozen"(#pLft_0, #p, #m, #frozen) ]]
normal

pred "<A as gilogic::prophecies::FrozenOwn<(u64,)>>::frozen_own"(+THIS, model, G_3):
  (#G_3 == {{ #x }})
  * (#THIS == {{ #x, #y }}) * (#model == #x) * (#y == ((#x) i+ (1i)))
  * (G_3 == #G_3)
  * (model == #model)
  * (THIS == #THIS);


pred a_ref_mut_x(+pLft_a, +this, model, frozen):
  "<A as gilogic::prophecies::FrozenOwn<(u64,)>>::mut_ref_own_frozen"(pLft_a, #this, #model, #frozen)
  * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj)
  * (#this == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }})
  * (frozen == #frozen)
  * (model == #model)
  * (this == #this);


pred "<A as gilogic::prophecies::Ownable>::own"(+self, model):
  (#self == {{ #x, #y }}) * (#model == #x) * (#y == ((#x) i+ (1i)))
  * (model == #model)
  * (self == #self);


pred "<A as gilogic::prophecies::FrozenOwn<(u64,)>>::mut_ref_own_frozen"(+pLft_0, +this, model, frozen):
  (#model == {{ #a, #b }})
  * "<A as gilogic::prophecies::FrozenOwn<(u64,)>>::mut_ref_inner_frozen"(pLft_0, #this, #frozen)
  * <value_observer>(l-nth(#this, 1i); #a)
  * (#lvar_0 == #b)
  * <pcy_value>(l-nth(#this, 1i); #lvar_0)
  * types(#lvar_1: Obj, #lvar_2: List, #lvar_3: Obj)
  * (#this == {{ {{ #lvar_1, #lvar_2 }}, #lvar_3 }})
  * (frozen == #frozen)
  * (model == #model)
  * (this == #this);


pred "<&mut A as gilogic::prophecies::Ownable>::own"(+pLft_erased, +self, model):
  "<A as gilogic::prophecies::FrozenOwn<()>>::mut_ref_own_frozen"(pLft_erased, #self, #model, {{  }})
  * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj)
  * (#self == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }})
  * (model == #model)
  * (self == #self);


pred "<A as gilogic::prophecies::FrozenOwn<(u64,)>>::mut_ref_inner_frozen"(+pLft_0, +this, frozen):
  <value>(l-nth(l-nth(#this, 0i), 0i), l-nth(l-nth(#this, 0i), 1i), {{ "adt", "A", {{  }} }}; #v)
  * <pcy_controller>(l-nth(#this, 1i); #repr)
  * "<A as gilogic::prophecies::FrozenOwn<(u64,)>>::frozen_own"(#v, #repr, #frozen)
  * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj)
  * (#this == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }})
  * (frozen == #frozen)
  * (this == #this);
guard: <lft>(pLft_0; true);

pred "<A as gilogic::prophecies::FrozenOwn<()>>::mut_ref_own_frozen"(+pLft_0, +this, model, frozen):
  (#model == {{ #a, #b }})
  * "<A as gilogic::prophecies::FrozenOwn<()>>::mut_ref_inner_frozen"(pLft_0, #this, #frozen)
  * <value_observer>(l-nth(#this, 1i); #a)
  * (#lvar_0 == #b)
  * <pcy_value>(l-nth(#this, 1i); #lvar_0)
  * types(#lvar_1: Obj, #lvar_2: List, #lvar_3: Obj)
  * (#this == {{ {{ #lvar_1, #lvar_2 }}, #lvar_3 }})
  * (frozen == #frozen)
  * (model == #model)
  * (this == #this);


pred "<A as gilogic::prophecies::FrozenOwn<()>>::mut_ref_inner_frozen"(+pLft_0, +this, frozen):
  <value>(l-nth(l-nth(#this, 0i), 0i), l-nth(l-nth(#this, 0i), 1i), {{ "adt", "A", {{  }} }}; #v)
  * <pcy_controller>(l-nth(#this, 1i); #repr)
  * "<A as gilogic::prophecies::FrozenOwn<()>>::frozen_own"(#v, #repr, #frozen)
  * types(#lvar_0: Obj, #lvar_1: List, #lvar_2: Obj)
  * (#this == {{ {{ #lvar_0, #lvar_1 }}, #lvar_2 }})
  * (frozen == #frozen)
  * (this == #this);
guard: <lft>(pLft_0; true);

pred "<A as gilogic::prophecies::FrozenOwn<()>>::frozen_own"(+this, repr, frozen):
  "<A as gilogic::prophecies::Ownable>::own"(#this, #repr) * (#frozen == {{  }})
  * (frozen == #frozen)
  * (repr == #repr)
  * (this == #this);


proc freeze_x(pLft_a, x) {
        m_1x := x;
        g_0 := m_1x;
        m_1x := [alloc]({{ "ref", true, {{ "adt", "A", {{  }} }} }});
        u := [store_value](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, {{ "adt", "A", {{  }} }} }}, g_0);
  bb0:  g_2 := [load_value](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, {{ "adt", "A", {{  }} }} }}, true);
        g_1 := l-nth(g_2, 0i);
        g_3 := [new_lft]();
        pLft_5 := l-nth(g_3, 0i);
        m_2 := g_1;
        g_4 := m_2;
        g_5 := "gilogic::prophecies::freeze_params::<(u64,), A>"(pLft_a, g_4);
        m_0 := g_5;
        goto bb1;
  bb1:  g_6 := m_0;
        ret := g_6;
        goto rlab;
  rlab: u := [free](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, {{ "adt", "A", {{  }} }} }});
        return
};

